os-server/src/main/java/app/openschool/category/CategoryServiceImpl.java

```diff
@@ -1,100 +1,179 @@
 package app.openschool.category;
 
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toList;
+
 import app.openschool.category.api.dto.CategoryDto;
+import app.openschool.category.api.dto.CreateOrModifyCategoryRequest;
+import app.openschool.category.api.dto.ParentAndSubCategoriesDto;
 import app.openschool.category.api.dto.PreferredCategoryDto;
+import app.openschool.category.api.exception.ImageNotExistsException;
+import app.openschool.category.api.exception.IncorrectCategoryTitleException;
 import app.openschool.category.api.mapper.CategoryMapper;
-import java.util.ArrayList;
-import java.util.HashMap;
+import app.openschool.common.services.aws.S3Service;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
 import java.util.List;
+import java.util.Locale;
 import java.util.Map;
-import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.MessageSource;
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.stereotype.Service;
-import org.springframework.transaction.annotation.Transactional;
+import org.springframework.web.multipart.MultipartFile;
 
 @Service
 public class CategoryServiceImpl implements CategoryService {
 
   private final CategoryRepository categoryRepository;
+  private final MessageSource messageSource;
+  private final S3Service s3Service;
+  private final Logger logger = LoggerFactory.getLogger("CategoryServiceImpl");
 
-  public CategoryServiceImpl(CategoryRepository categoryRepository) {
+  public CategoryServiceImpl(
+      CategoryRepository categoryRepository, MessageSource messageSource, S3Service s3Service) {
     this.categoryRepository = categoryRepository;
+    this.messageSource = messageSource;
+    this.s3Service = s3Service;
   }
 
   @Override
-  public Page<CategoryDto> findAllCategories(Pageable pageable) {
-    return CategoryMapper.toCategoryDtoPage(categoryRepository.findAllCategories(pageable));
+  public Page<CategoryDto> findAllParentCategories(Pageable pageable) {
+    return CategoryMapper.toCategoryDtoPage(
+        categoryRepository.findByParentCategoryIsNull(pageable));
   }
 
   @Override
-  @Transactional
   public Map<String, List<PreferredCategoryDto>> findCategoriesByTitle(String title) {
 
-    Map<String, List<PreferredCategoryDto>> mappedCategories;
+    Map<String, List<PreferredCategoryDto>> result;
 
     if (StringUtils.isBlank(title)) {
-      return mapCategoriesToSubcategories(
-          categoryRepository.findCategoriesByParentCategoryId(null));
+      result = findBySubCategories(categoryRepository.findByParentCategoryIsNotNull());
```

Comment:
why do you first assign the returned value to the map, then doing putAll the same result to the same map